lib/rouge/lexers/yang.rb

```diff
@@ -0,0 +1,147 @@
+      def self.top_stmts_keywords
+        @top_stms_keywords ||= Set.new %w(
+          module submodule
+        )
+      end
+
+      def self.module_header_stmts_keywords
+        @module_header_stmts_keywords ||= Set.new %w(
+          yang-version namespace prefix belongs-to
+        )
+      end
+
+      def self.meta_stmts_keywords
+        @meta_stmts_keywords ||= Set.new %w(
+          organization contact description reference revision
+        )
+      end
+
+      def self.linkage_stmts_keywords
+        @linkage_stmts_keywords ||= Set.new %w(
+          import include revision-date
+        )
+      end
+
+      def self.body_stmts_keywords
+        @body_stms_keywords ||= Set.new %w(
+          extension feature identity typedef grouping augment rpc notification
+          deviation action argument identity if-feature input output
+        )
+      end
+
+      def self.data_def_stmts_keywords
+        @data_def_stms_keywords ||= Set.new %w(
+          container leaf-list leaf list choice anydata anyxml uses case config
+          deviate must when presence refine
+        )
+      end
+
+      def self.type_stmts_keywords
+        @type_stmts_keywords ||= Set.new %w(
+          type units default status bit enum error-app-tag error-message
+          fraction-digits length min-elements max-elements modifier ordered-by
+          path pattern position range require-instance value yin-element base
+        )
+      end
+
+      def self.list_stmts_keywords
+        @list_stmts_keywords ||= Set.new %w(
+          key mandatory unique
+        )
+      end
+
+      #RFC7950 other keywords
+      def self.constants_keywords
+        @constants_keywords ||= Set.new %w(
+          true false current obsolete deprecated add delete replace
+          not-supported invert-match max min unbounded user
+        )
+      end
+
+      #RFC7950 Built-In Types
+      def self.types
+        @types ||= Set.new %w(
+          binary bits boolean decimal64 empty enumeration int8 int16 int32
+          int64 string uint8 uint16 uint32 uint64 union leafref identityref
+          instance-identifier
+        )
+      end
+
+      state :comment do
+        rule %r/[^*\/]/, Comment
+        rule %r/\/\*/, Comment, :comment
+        rule %r/\*\//, Comment, :pop!
+        rule %r/[*\/]/, Comment
+      end
+
+      #Keyword::Reserved
+      #groups Name::Tag, Text::Whitespace
+      state :root do
+        rule %r/\s+/, Text::Whitespace
+        rule %r/[\{\}\;]+/, Punctuation
+        rule %r/(?<![\-\w])(and|or|not|\+|\.)(?![\-\w])/, Operator
+
+        rule %r/"(?:\\"|[^"])*?"/, Str::Double #for double quotes
+        rule %r/'(?:\\'|[^'])*?'/, Str::Single #for single quotes
+
+        rule %r/\/\*/, Comment, :comment
+        rule %r/\/\/.*?$/, Comment
+
+        #match BNF stmt for `node-identifier` with [ prefix ":"]
+        rule %r/(?:^|(?<=[\s\{\}\;]))([^;{}\s\*\+\'\"\:\/]+)(:)([^;{}\s\*\+\'\"\:\/]+)(?=[\s\{\}\;])/ do
```

Comment:
I'm a bit confused by this rule. What is `^|(?<=[\s\{\}\;])` meant to match at the beginning of the regex?

Reply:
Separately, is the lookahead at the end of the regex necessary? It's a non-goal of Rouge to check for syntax correctness and it would aid maintainability to keep things minimal.

Reply:
IMO, this rule is the most complicated. I tried it and couldn't find any optimization.
At following combinations, the word `namespace-prefix` have to be matched with token `Name::Namespace`:
1. (BEGIN OF LINE)**namespace-prefix**:example{ (END OF LINE)
2. (BEGIN OF LINE)blabla asdd **namespace-prefix**:example(END OF LINE) 

At following combinations, the word `namespace-prefix` have **NOT** to be matched with token `Name::Namespace`, because it isn't a namespace-definition in YANG:
1. (BEGIN OF LINE)namespace-prefix:example:only-string-no-namespace { (END OF LINE)
2. (BEGIN OF LINE)I am only string for namespace-prefix:example:only-string-no-namespace{ (END OF LINE)

Maybe you could follow and find any optimization for this rule?

Reply:
Got it. Without necessarily changing the regex structure, could we clean it up a little? Would the below work?

```suggestion
        rule %r/(?:^|(?<=[\s{};]))([\w-]+)(:)([\w-]+)(?=[\s{};])/ do
```

Reply:
Thanks! I have checked it out and found in example file following string`type yang:xpath1.0;`.
Therefore I added literal dot into rule. 
New rule:
`%r/(?:^|(?<=[\s{};]))([\w\-\.]+)(:)([\w\-\.]+)(?=[\s{};])/`